include/ghex/communication_object_ipr.hpp

```diff
@@ -0,0 +1,564 @@
+                    using handle_type = communication_handle_ipr<communicator_type,grid_type,domain_id_type>;
+                    using pattern_type = pattern<communicator_type,grid_type,domain_id_type>;
+                    using pattern_container_type = pattern_container<communicator_type,grid_type,domain_id_type>;
+
+                    template<typename D, typename F>
+                    using buffer_info_type = buffer_info<pattern_type,D,F>;
+
+                private:
+
+                    // member types
+
+                    using address_type = typename communicator_type::address_type;
+                    using index_container_type = typename pattern_type::index_container_type;
+                    using pack_function_type = std::function<void(void*, const index_container_type&, void*)>; // last argument intended as CUDA stream
+
+                    /** @brief pair of domain ids + tag id with ordering */
+                    struct domain_id_pair_and_tag {
+
+                            domain_id_type first_id;
+                            domain_id_type second_id;
+                            int tag;
+
+                            bool operator<(const domain_id_pair_and_tag& other) const noexcept {
+                                return (first_id < other.first_id ? true :
+                                    (first_id > other.first_id ? false :
+                                        (second_id < other.second_id ? true :
+                                            (second_id > other.second_id ? false :
+                                                (tag < other.tag)))));
+                            }
+
+                    };
+
+                    /** @brief Holds a pointer to a set of iteration spaces and a callback function pointer
+                     * which is used to store a field's pack or unpack member function.
+                     * This class also stores the offset in the serialized buffer in bytes.
+                     * The type-erased field_ptr member is only used for the gpu-vector-interface. // TO DO: verify is still needed in this case
+                     * @tparam Function either pack or unpack function pointer type */
+                    template<typename Function>
+                    struct field_info {
+                            using index_container_type = typename pattern_type::map_type::mapped_type;
+                            Function call_back;
+                            const index_container_type* index_container;
+                            std::size_t offset;
+                            void* field_ptr;
+                    };
+
+                    /** @brief Holds serial buffer memory and meta information associated with it.
+                     * @tparam Vector contiguous buffer memory type
+                     * @tparam Function pack function pointer type (not used for unpacking)*/
+                    template<class Vector, class Function>
```

Comment:
This struct also is the same as co_2. Maybe instead of composing classes, we can put these sub-structs into an impl namespace in a separate file and use them when we need them

Reply:
Moved to refactoring issue.